File: RunARMBenchmark.py
from pyspark.sql import SparkSession
from pyspark.sql.functions import col
import time

# Create Spark session
spark = SparkSession.builder \
    .appName("ARM Benchmark Python") \
    .master("local[*]") \
    .config("spark.sql.shuffle.partitions", "200") \
    .config("spark.serializer", "org.apache.spark.serializer.KryoSerializer") \
    .getOrCreate()

def benchmark(name):
    """Decorator function to time operations"""
    def decorator(func):
        def wrapper(*args, **kwargs):
            start = time.time()
            result = func(*args, **kwargs)
            end = time.time()
            print(f"{name} took {end - start:.9f} seconds")
            return result
        return wrapper
    return decorator

print("Starting ARM Performance Benchmark Suite")
print(f"Spark version: {spark.version}")
print(f"Python version: {spark.sparkContext.pythonVer}")
print(f"Available cores: {spark.sparkContext.defaultParallelism}")

# (a) Transformation Benchmark
print("\n" + "=" * 80)
print("(a) Transformation Benchmark")
print("=" * 80)

@benchmark("Python Transformation")
def transformation_benchmark():
    data = spark.range(1, 100000000)
    result = data.filter(col("id") % 2 == 0).count()
    print(f"Count: {result}")
    return result

transformation_benchmark()

# (b) SQL Query Benchmark
print("\n" + "=" * 80)
print("(b) SQL Query Benchmark")
print("=" * 80)

@benchmark("Python SQL Query")
def sql_query_benchmark():
    df = spark.range(1, 50000000).withColumn("val", col("id") * 2)
    df.createOrReplaceTempView("t")
    spark.sql("SELECT avg(val) FROM t").show()

sql_query_benchmark()

# (c) Parquet I/O Benchmark
print("\n" + "=" * 80)
print("(c) Parquet I/O Benchmark")
print("=" * 80)

@benchmark("Python Parquet I/O")
def parquet_io_benchmark():
    df = spark.range(1, 10000000).withColumn("val", col("id") * 3)
    df.write.mode("overwrite").parquet("test_parquet_python")
    read_df = spark.read.parquet("test_parquet_python")
    count = read_df.count()
    print(f"Parquet Count: {count}")
    return count

parquet_io_benchmark()

print("\n" + "=" * 80)
print("Benchmark Suite Completed Successfully")
print("=" * 80)

spark.stop()

File: RunARMBenchmark.scala
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._

val spark = SparkSession.builder()
  .appName("ARM Benchmark")
  .master("local[*]")
  .config("spark.sql.shuffle.partitions", "200")
  .config("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
  .getOrCreate()

def benchmark(name: String)(f: => Unit): Unit = {
  val start = System.nanoTime()
  f
  val end = System.nanoTime()
  println(s"$name took ${(end - start) / 1e9} seconds")
}

println("Starting ARM Performance Benchmark Suite")
println(s"Spark version: ${spark.version}")
println(s"Java version: ${System.getProperty("java.version")}")
println(s"Available cores: ${Runtime.getRuntime.availableProcessors}")

// (a) Transformation Benchmark
println("\n" + "=" * 80)
println("(a) Transformation Benchmark")
println("=" * 80)

benchmark("Scala Transformation") {
  val data = spark.range(1, 100000000)
  val result = data.filter(col("id") % 2 === 0).count()
  println(s"Count: $result")
}

// (b) SQL Query Benchmark
println("\n" + "=" * 80)
println("(b) SQL Query Benchmark")
println("=" * 80)

benchmark("Scala SQL Query") {
  val df = spark.range(1, 50000000).withColumn("val", col("id") * 2)
  df.createOrReplaceTempView("t")
  spark.sql("SELECT avg(val) FROM t").show()
}

// (c) Parquet I/O Benchmark
println("\n" + "=" * 80)
println("(c) Parquet I/O Benchmark")
println("=" * 80)

benchmark("Scala Parquet I/O") {
  val df = spark.range(1, 10000000).withColumn("val", col("id") * 3)
  df.write.mode("overwrite").parquet("test_parquet_scala")
  val readDf = spark.read.parquet("test_parquet_scala")
  println(s"Parquet Count: ${readDf.count()}")
}

println("\n" + "=" * 80)
println("Benchmark Suite Completed Successfully")
println("=" * 80)

spark.stop()